File: c++/leetcode/83. Remove Duplicates from Sorted List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
  ListNode *deleteDuplicates(ListNode *head)
  {
    if (!head || !head->next)
      return head;

    ListNode *slow = head;
    ListNode *fast = head->next;
    slow->next = nullptr;

    while (fast)
    {
      if (fast->val != slow->val)
      {
        slow->next = fast;
        slow = fast;
        fast = fast->next;
        slow->next = nullptr;
      }
      else
        fast = fast->next;
    }

    return head;
  }
};

File: python/leetcode/83. Remove Duplicates from Sorted List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head

        slow = head
        fast = slow.next
        slow.next = None

        while fast:
            if fast.val != slow.val:
                slow.next = fast
                slow = fast
                fast = fast.next
                slow.next = None
            else:
                fast = fast.next

        return head
